### PR TITLE
Updates for GAMER SR fields

### PR DESCRIPTION
**Pull request by jzuhone**
## PR Summary

The GAMER special relativistic hydrodynamics feature is nearly merged into the main branch of GAMER. A number of changes have taken place in the code which motivates an update to the frontend here. 

* The `Temp` field is no longer the dimensionless temperature kT/mc^2 but is now simply the temperature in Kelvin. This field is used extensively in calculations in its previous dimensionless form, so I have created the `temp_fraction` field to represent it. 
* We no longer need to calculate the reduced enthalpy in yt itself as GAMER will now output it to the dataset file. This means we can eliminate the code that calculates the reduced enthalpy, which is done here.
* I have also eliminated the code that calculates fields for the gamma-law EOS, as supporting it is not planned in GAMER. This simplifies the code considerably. 
* Some fields, such as 3-velocities, Lorentz factor, sound speed, or mach number may be stored in the file. If they are, yt will use them instead of calculating them. 
* I have reduced the number of fields which are calculated in Cython, as these calculations have been greatly simplified in some cases. The most expensive fields to calculate remain. 
* Though not related to SRHD, I added field aliases for `velocity_divergence` and `magnetic_field_divergence` as these may be set by the user to appear in the dataset. 

The answer tests will fail and will need to be updated. I will do some testing to ensure that the calculations are correct before we merge. 

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by matthewturk** [comment]
Thanks for keeping on top of this @jzuhone !

**Comment by jzuhone** [comment]
Update--the nose test failures here are expected. I have checked by hand that the differences all arise from:

a. replacing `amu` with `mh` in the calculation of temperature/pressure (depending on which direction one has going)
b. that the new GAMER outputs define the `Temp` value for SRHD in terms of Kelvin (the same as non-SR HD) instead of kT/mc^2. 

I will need to replace the `JetICMWall` dataset on `yt-project.org/data`, then we'll need to replace it on Jenkins, then I'll need to generate new answers. In the meantime, can @hyschive take a look at these changes? 

The Python 3.12 failures appear unrelated (^^ @neutrinoceros)

**Comment by jzuhone** [comment]
@hyschive want to have a look at this before I update the test answers?

**Comment by hyschive** [comment]
> @hyschive want to have a look at this before I update the test answers?
@jzuhone Yes. I'll finish the review in the next few days. Let's update the test answers after that.

File: yt/frontends/gamer/fields.py [hunk 1]
```diff
@@ -59,42 +69,110 @@ def setup_fluid_fields(self):
         unit_system = self.ds.unit_system
         unit_system.registry = self.ds.unit_registry  # TODO: Why do I need this?!
 
+        temp_conv = pc.kb / (self.ds.mu * pc.mh)
```
Comment (hyschive) [inline yt/frontends/gamer/fields.py:72]: How about setting `temp_conv = 1 / mu` for dimensionless simulations to be more consistent with GAMER (e.g., see [here](https://github.com/gamer-project/gamer/blob/main/src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp#L56-L57))? So something like
```
        if self.ds.unit:
            temp_conv = pc.kb / (self.ds.mu * pc.mh)
        else:
            temp_conv = self.ds.arr(1.0, "code_velocity**2/code_temperature") / self.ds.mu
```
where `self.ds.unit` is set in `data_structures.py` (see the comment therein). Also check the following comments on other physical constants.

File: yt/frontends/gamer/fields.py [hunk 2]
```diff
@@ -59,42 +69,110 @@ def setup_fluid_fields(self):
...
+
         if self.ds.srhd:
             c2 = pc.clight * pc.clight
+            invc2 = 1.0 / c2
```
Comment (hyschive) [inline yt/frontends/gamer/fields.py:76]: Similarly, how about setting all physical constants to unity for dimensionless simulations? For example,
```
            if self.ds.unit:
                c2 = pc.clight * pc.clight
                invc2 = 1.0 / c2
                c = pc.clight.in_units("code_length / code_time")
            else:
                c = self.ds.arr(1.0, "code_velocity")
                c2 = c*c
                invc2 = 1.0 / c2
```
Together with `temp_conv`, they ensure no extra unit conversion in dimensionless simulations. I've confirmed that these modifications are necessary to get the correct results when computing, for example, the effective adiabatic index `("gas", "gamma")` in the `Riemann` test problem.

File: yt/frontends/gamer/data_structures.py [hunk 3]
```diff
@@ -367,6 +367,8 @@ def _parse_parameter_file(self):
         else:
             self.mhd = 0
             self.srhd = 0
+            # set dummy value of mu here to avoid more complicated workarounds later
+            self.mu = 1.0
```
Comment (hyschive) [inline yt/frontends/gamer/data_structures.py:371]: Replace `self.parameters.setdefault("Opt__Unit", 0)` on line 374 with `self.unit = parameters.get("Opt__Unit", 0)`. Also check the related comments in `fields.py`.

File: doc/source/examining/loading_data.rst
Comment (hyschive) [inline doc/source/examining/loading_data.rst:1]: I don't think `four_velocity_t` is currently supported.

**Review by hyschive — CHANGES_REQUESTED** [review]
@jzuhone The PR looks good in general. I've put some inline comments, including a few bug fixes. In addition, here are two minor comments, which are just for the records and do not need to be addressed in this PR.
- The speed-of-light `c` conversion is a bit confusing to me. IIUC, for example, `cfields.pyx` has implicitly assumed that the specific reduced enthalpy `h` is always normalized to `c` (regardless of the code velocity unit), while `fields.py` has not made such an assumption. This is fine since GAMER always normalizes to `c` for SRHD. But I just want to make sure that I understand it correctly.
- Some physical constants have slightly different values in unyt and gamer. For example, the neutral atomic hydrogen mass `mh` is `1.6737352238051867e-24` and `1.6735327579879997e-24` in unyt and gamer, respectively. The Boltzmann constant `kB` in erg/K is `1.3806488e-16` and `1.38064852e-16` in unyt and gamer, respectively. These will cause some small discrepancies between the values output by gamer directly and post-processed by yt (e.g., `('gamer','Pres')` and `('gas','pressure')`).

File: yt/frontends/gamer/cfields.pyx [hunk 4]
```diff
@@ -117,156 +66,138 @@ cdef class SRHDFields:
...
-            lf = self._lorentz_factor(rho[i], mx[i], my[i], mz[i], kT[i])
-            outp[i] = rho[i] / lf * self._c2 * kT[i]
+            outp[i] = self._four_vel(rho[i], mi[i], h[i])
         return out
 
     @cython.boundscheck(False)
     @cython.wraparound(False)
     @cython.cdivision(True)
-    def specific_thermal_energy(self, dens, momx, momy, momz, temp):
+    def specific_thermal_energy(self, dens, momx, momy, momz, temp, enth):
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/cfields.pyx:144]: Thanks, you are right.

File: doc/source/examining/loading_data.rst [hunk 5]
```diff
@@ -1868,12 +1868,20 @@ are supported, and the following fields are defined:
 * ``("gas", "four_velocity_[txyz]")``: Four-velocity fields :math:`U_t, U_x, U_y, U_z`
 * ``("gas", "lorentz_factor")``: Lorentz factor :math:`\gamma = \sqrt{1+U_iU^i/c^2}`
   (where :math:`i` runs over the spatial indices)
+* ``("gas", "specific_reduced_enthalpy")``: Specific reduced enthalpy :math:`h = \epsilon + p/\rho`
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply doc/source/examining/loading_data.rst:1871]: Done.

File: yt/frontends/gamer/cfields.pyx [hunk 6]
```diff
@@ -117,156 +66,138 @@ cdef class SRHDFields:
...
         cdef np.float64_t[:] outp = out.ravel()
         cdef np.float64_t cs
         cdef int i
 
         for i in range(outp.shape[0]):
-            cs = self.cs(kT[i], self._c, self._gamma)
+            cs = self._cs(kT[i], h[i])
             us = cs / math.sqrt(1.0 - cs**2 / self._c2)
-            u = math.sqrt(mx[i]**2 + my[i]**2 + mz[i]**2) / (rho[i] * (self.h(kT[i], self._gamma) + 1.0))
+            u = math.sqrt(mx[i]**2 + my[i]**2 + mz[i]**2) / (rho[i] * (h[i] + 1.0))
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/cfields.pyx:201]: That does result in a speedup, thanks for noticing it.

File: yt/frontends/gamer/fields.py [hunk 7]
```diff
@@ -25,10 +26,19 @@ class GAMERFieldInfo(FieldInfoContainer):
         ("Engy", (erg_units, ["total_energy_density"], None)),
         ("CRay", (erg_units, ["cosmic_ray_energy_density"], None)),
         ("Pote", (pot_units, ["gravitational_potential"], None)),
+        ("Temp", ("code_temperature", ["temperature"], None)),
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:29]: That is correct that Kelvin is the default.

File: yt/frontends/gamer/fields.py [hunk 8]
```diff
@@ -25,10 +26,19 @@ class GAMERFieldInfo(FieldInfoContainer):
...
+        ("Enth", (pot_units, ["specific_reduced_enthalpy"], None)),
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:30]: So the way this currently works for SRHD datasets, if `VelX/Y/Z` are found in the dataset, yt will use them to define `velocity_x/y/z`. I'm going to set this up for Newtonian HD also. 

I'll do the same for pressure.

File: doc/source/examining/loading_data.rst
Comment (jzuhone) [reply doc/source/examining/loading_data.rst:1]: @hyschive this field comes for free if `lorentz_factor` is defined:

https://github.com/yt-project/yt/blob/d5b33354efacc821a859af097751379e8121f5cc/yt/fields/astro_fields.py#L156-L165

GAMER defines `lorentz_factor` from its own fields (without catastrophic cancellation) and this is sufficient to give you `four_velocity_t`.

File: yt/frontends/gamer/cfields.pyx [hunk 9]
```diff
@@ -117,156 +66,138 @@ cdef class SRHDFields:
...
         cdef int i
 
         for i in range(outp.shape[0]):
-            hp = self.h(kT[i], self._gamma) + 1.0
             ux = self._four_vel(rho[i], kT[i], mx[i])
             uy = self._four_vel(rho[i], kT[i], my[i])
             uz = self._four_vel(rho[i], kT[i], mz[i])
             u2 = ux**2 + uy**2 + uz**2
-            lf = self._lorentz_factor(rho[i], mx[i], my[i], mz[i], kT[i])
+            lf = self._lorentz_factor(rho[i], mx[i], my[i], mz[i], h[i])
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/cfields.pyx:176]: Thanks for catching that.

File: yt/frontends/gamer/cfields.pyx [hunk 10]
```diff
@@ -117,156 +66,138 @@ cdef class SRHDFields:
...
+        cdef np.float64_t[:] h = enth.ravel()
 
         out = np.empty_like(dens)
         cdef np.float64_t[:] outp = out.ravel()
         cdef np.float64_t cs
         cdef int i
 
         for i in range(outp.shape[0]):
-            cs = self.cs(kT[i], self._c, self._gamma)
+            cs = self._cs(kT[i], h[i])
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/cfields.pyx:199]: Thanks for catching that.

File: yt/frontends/gamer/fields.py [hunk 11]
```diff
@@ -25,10 +26,19 @@ class GAMERFieldInfo(FieldInfoContainer):
...
+        ("Mach", ("dimensionless", ["mach_number"], None)),
+        ("Cs", (vel_units, ["sound_speed"], None)),
+        ("DivVel", ("1/code_time", ["velocity_divergence"], None)),
         # MHD fields on disk (CC=cell-centered)
         ("CCMagX", (b_units, [], "B_x")),
         ("CCMagY", (b_units, [], "B_y")),
         ("CCMagZ", (b_units, [], "B_z")),
+        (
+            "DivMag",
+            ("code_magnetic/code_length", ["magnetic_field_divergence"], None),
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:40]: I will remove this then.

File: yt/frontends/gamer/fields.py [hunk 12]
```diff
@@ -59,42 +69,110 @@ def setup_fluid_fields(self):
...
+            # total energy density
+            def _total_energy_density(field, data):
+                return data["gamer", "Engy"] + data["gamer", "Dens"] * c2
+
+            self.add_field(
+                ("gas", "total_energy_density"),
+                sampling_type="cell",
+                function=_total_energy_density,
+                units=unit_system["pressure"],
+            )
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:160]: I've removed the alias to `total_energy_density` in the `known_other_fields` so we can set it to `reduced_total_energy_density` or `total_energy_density` depending on SRHD or HD.

File: yt/frontends/gamer/fields.py [hunk 13]
```diff
@@ -270,6 +313,12 @@ def _mach_number(field, data):
                 ("gas", "density"), ("gamer", "Dens"), units=unit_system["density"]
             )
 
+            self.alias(
+                ("gas", "total_energy_density"),
+                ("gamer", "Engy"),
+                units=unit_system["pressure"],
+            )
```
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:320]: I've removed the alias to `total_energy_density` in the `known_other_fields` so we can set it to `reduced_total_energy_density` or `total_energy_density` depending on SRHD or HD.

File: yt/frontends/gamer/cfields.pyx
[the remark this answers is not in the document]
Comment (jzuhone) [reply yt/frontends/gamer/cfields.pyx:1]: Thanks for the fix.

**Comment by jzuhone** [comment]
@hyschive:

> The speed-of-light c conversion is a bit confusing to me. IIUC, for example, cfields.pyx has implicitly assumed that the specific reduced enthalpy h is always normalized to c (regardless of the code velocity unit), while fields.py has not made such an assumption. This is fine since GAMER always normalizes to c for SRHD. But I just want to make sure that I understand it correctly.

I was thinking about this myself--of course the attributes `self._c` and `self._c2` are equal to unity in the class `SRHDFields` in `cfields.pyx`.

If we have no reason to think that the `code_velocity` from SRHD datasets will be anything other than `c`, we can just remove all references to `c` in `cfields.pyx`.

File: yt/frontends/gamer/fields.py (hunk 1 above)
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:69]: Done.

File: yt/frontends/gamer/data_structures.py (hunk 3 above)
Comment (jzuhone) [reply yt/frontends/gamer/data_structures.py:371]: Done.

**Comment by hyschive** [comment]
> If we have no reason to think that the `code_velocity` from SRHD datasets will be anything other than `c`, we can just remove all references to `c` in `cfields.pyx`.

@jzuhone I cannot think of a situation in which we prefer not to use `c` as the basic code unit. So if you agree, let's just remove all `c` in `cfields.pyx`. But we may still add a sanity check somewhere in case that assumption is ever broken.

File: yt/frontends/gamer/fields.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hyschive) [reply yt/frontends/gamer/fields.py]: @jzuhone Would it be possible to still keep it but make it dimensionless?

**Comment by jzuhone** [comment]
@hyschive then I think I have addressed everything in my recent commits.

File: yt/frontends/gamer/fields.py (hunk 2 above)
Comment (jzuhone) [reply yt/frontends/gamer/fields.py:81]: Done.

**Review by hyschive — DISMISSED** [review]

**Comment by jzuhone** [comment]
@yt-fido test this please

**Comment by jzuhone** [comment]
@yt-fido test this please

**Comment by jzuhone** [comment]
@Xarthisius a little confused about the Jenkins failure here, it seems to be persistent but unrelated

**Comment by jzuhone** [comment]
@yt-fido test this please

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Review by hyschive — APPROVED** [review]

**Comment by jzuhone** [comment]
We updated the dataset used on the build box (JetICMWall) for those tests so that's the reason. The fact that it coincides with this commit is (somewhat of) a coincidence.

@hyschive, the head GAMER developer, has already approved PR #4782, so it would be helpful to have a second approval for just code style, etc on it.

**Pull request merged**